tool/net/redbean.c

```diff
@@ -3720,6 +3725,41 @@ static int LuaFetch(lua_State *L) {
     maxredirects = luaL_optinteger(L, -1, maxredirects);
     lua_getfield(L, 2, "numredirects");
     numredirects = luaL_optinteger(L, -1, numredirects);
+    lua_getfield(L, 2, "headers");
+    if (!lua_isnil(L, -1)) {
+      if (!lua_istable(L, -1))
+        return luaL_argerror(L, 2, "invalid headers value; table expected");
+
+      lua_pushnil(L);
+      while (lua_next(L, -2)) {
+        if (lua_type(L, -2) == LUA_TSTRING) {  // skip any non-string keys
+          key = lua_tolstring(L, -2, &keylen);
+          if (!IsValidHttpToken(key, keylen))
+            return luaL_argerror(L, 2, "invalid header name");
+
+          val = lua_tolstring(L, -1, &vallen);
+          if (!(hdr = gc(EncodeHttpHeaderValue(val, vallen, 0))))
+            return luaL_argerror(L, 2, "invalid header value encoding");
+
+          // Content-Length and Connection will be overwritten;
+          // skip them to avoid duplicates
+          if (strcasecmp(key, "content-length") &&
```

Comment:
The high performance redbean way to do this, using perfect hash tables, is:

```c
int h;
if ((h = GetHttpHeader(key, keylen)) != -1 &&
    h == kHttpContentLength) {
  // ...
}
```

One thing this allows is, if the HTTP header is hard-coded (i.e. GetHttpHeader returns non -1) then you can round-trip it through `GetHttpHeaderName(h)` if you want to normalize the case, e.g. turning `"content-type"` into `"Content-Type"`.

Reply:
I considered this, but was under the impression that `GetHttpHeader(key, keylen)` will check against the request buffer, which is not the case here, as it's a header provided by user to be sent as part of a Fetch request. Am I wrong?

Reply:
No `GetHttpHeader` is a pure function. It doesn't reference any globals.

Reply:
Right; updated.